Sprite Status 3 Custom Graphics Handler

SubconsciousEye opened the pull request:
This was intended to be made earlier, but IRL events pushed it some weeks.

## Preamble
In Vanilla SMW, Statuses 2 and 5 (Falling and Lava Deaths respectively) both run a subroutine that handles death graphics for the given sprite. A special condition at the start of the routine checks for bit 0 of the !167A table (AND #$01), and if it's set, it runs the sprite's Main (Status 8) routine to handle graphics instead of the generic routine. This very behavior is also what allows custom sprites to show proper GFX in the case it ends up dying.

Sprite Status 3 (Smushed) does not have this aforementioned check, so PIXI (and maybe older Sprite Tools, haven't checked too much) has a special case for custom sprites. This special case essentially just runs Sprite Main after running Status 3 code for Custom Sprites, regardless of the unique !extra_prop_2 status override property (specifically bit 6, as bit 7 will always call the Main Status if it's set).

Normally, the enforced "Call Sprite Main" code works well for custom sprites, but there is one minor oversight where !14C8 is zeroed in the Status 3 routine and be desynced from its stack backup, which is used for the Status check to call the Sprite's Main routine. This oversight can cause some unintended side effects if a sprite doesn't take it into account or expect it.

What this Pull Request does is implement a Custom Graphics Handler for Status 3 utilizing !167A, and hopefully deprecates the old system.

## Pre-Testing Notes
There are some sprites in SMW, such as the P-Switch and Rex, that don't actually use Status 3, but instead "spoof" it through their main code. Interestingly, the P-Switch *does* call Status 3's Generic GFX Handler, but doesn't actually use said status under normal circumstances. Given these cases and how some Custom Sprites also utilize this same tactic, such sprites are not included in testing.

SMW's Koopas (Sprite 04-07) have an unusual interaction where if the player lands on a Koopa while in cape flight, the Koopa will become "Squished" and show garbage graphics. Technically, these graphics are the default for any sprite that isn't checked, but a Custom JSON that acts as a "Tweaked" Koopa is included for ease of testing.
Additionally, the Generic Graphics Handler for the Squished Status contains a check for SMW's Goomba (Sprite 0F), so a Custom JSON that acts as a "Tweaked" Goomba is included in the ZIP.

Both SMW's Dino Torch (Sprite 6F) and Sliding Beach Koopa (Sprite BD) *do* set bit 0 of !167A for death graphics, but still use some of the original SMW Status 3 code for getting graphics. As a result, these two sprites needed to be fixed in order to properly show their smushed pose when needed. The Dino Torch in particular uses a 16x16 for its squashed pose.

I created a special "Test" sprite to test the different methods of Custom Squished Graphics, both the old PIXI method and this new approach. Setting the Extra Bit for this Test Sprite will make it use a Fail-Safe for the old method, and clearing the Extra Bit will not utilize this Fail-Safe (thus "resetting" the sprite in the old method). Simply just walk up to the sprite to "squish" it.

I wasn't sure where to place the new code within main.asm, so I opted to slot them near the end of common hijacks/hexedits and near the end of most custom PIXI code. Additionally, I kept the formatting consistent with 90% of the other codes inside main.asm. Feel free to let me know or change as you see fit.

## Testing
Since there are so few sprites (both custom and vanilla) that can actually be put into Status 3 under normal circumstances, this test isn't very comprehensive. However, seeing as this is a test to ensure that these sprites can run properly without any additional modifications, this shouldn't affect results too much. Only (Fast) LoROM has been tested thus far, but it should work well with SA-1 assuming little-to-no edits to the ASM.

The list of tested sprites are as follows:
- SMW's Beach Koopa (00-03)
- SMW's Dino Torch (6F)
- SMW's Sliding Beach Koopa (BD)
- Tweaked SMW Koopa (04, different palette)
- Tweaked SMW Goomba (0F, different palette)
- [Thomas's Beach Koopa Disassembly](https://smwc.me/s/13892)
- [yoshicookiezeus's Sliding Koopa Disassembly](https://smwc.me/s/3487) (slight issue with gfx code, but otherwise runs)
- [Sonikku's SMB3 Goomba Family](https://smwc.me/s/23199) (only the non-winged variant)
- [Thomas's Sleepy Goomba](https://smwc.me/s/24079)
- Custom Status 3 Test Sprite

The Tweaked and Test Sprites can be found in this [ZIP](https://github.com/JackTheSpades/SpriteToolSuperDelux/files/13047696/Status3TestingSprites.zip), just insert them into any custom sprite slot.

This test was done with both the old PIXI method, and this new proposed method. No adverse effects were found when utilizing the new approach, as every sprite ran their smushed status without any hindrance.

## Alternate Approaches
Prior to the current approach this Pull Request is based on, I had considered other methodologies. I didn't opt for these alternatives since one would require Custom Sprites to update/fix their Acts Like, while another involves a lot of stack shenanigans to fix the affected Vanilla Sprites.

I believe this current implementation offers a nice middle-ground in terms of readability and code-size while also ensuring minimal edits and changes done to both sides. Please let me know if you believe one of the other alternative approaches better suits PIXI or if any (additional) concerns should be addressed.

Atari2 commented:
First of all, thanks for the detailed explanation and the testing, I really appreciate it when people take the time to do things properly. I will admit that even after reading everything 3 times I'm a bit confused as to what problem this PR is trying to fix, because it seems like the current way works "fine" except for some "unintended side effects" which is unclear to me as to what exactly these side-effects are, and as such I'm a bit confused. The code changes themselves look fairly simple and innocent, I am a bit worried about possiblities of these new hijacks conflicting with patches on SMWC, maybe that should be checked but I don't know how easy that is and I'm not comfortable asking you to do such a task so maybe I'll do it but it'll take a few days. 

If you could maybe bring an example of a sprite that breaks with the current behavior and how this patch fixes it, I could grasp the issue a bit better, in the meantime, I'll look for conflicting hijacks.

SubconsciousEye commented:
Admittedly, my thought process may have been unusual when I conceived this, so I apologize for not properly explaining my reasoning, and I will attempt to address this as best as I can.

I initially encountered this when I was working on an enhanced version of a Vanilla Sprite (technically multiple sprites in one sprite). The way I had the Death Poses set up was that I checked for !14C8,x being 3 (Smushed Status), and setting !1602,x to a value (let's say 9), otherwise set it to 8 (which would be Falling and Lava Death Poses). This went well, aside from the one frame where !14C8 is 0 coming from the usual Smushed Status Physics stuff, resulting in a wrong pose for that single frame. I fixed this by just terminating the sprite if !14C8 happens to be 0.

This, incidentally, is what made me investigate a bit deeper. A snippet from Thomas's all.log shows the following:
```
HandleSprSmushed:
	LDA $9D
	BNE CODE_019AFE
	LDA.w $1540,X
	BNE ShowSmushedGfx
	STZ.w $14C8,X
	RTS
```
Based on the code above, if the general timer (equivalent to !1540 in this case) is not zero, then branch to just after the RTS, otherwise zero !14C8 and move on to the next sprite to process. So far, the logic is all well and good, it makes sense to not do anything else afterwards, so it would stand to reason that PIXI, too, shouldn't do anything else. Unfortunately, since PIXI uses a stack backup of !14C8, a desync occurs, and the Sprite's Main code runs after its Squished Status.

It's understandable how such an oversight is present, as it would require a specific set of circumstances for something to be slightly off. That did get me thinking, however, on what'd happen if a sprite *didn't* properly account for such edge case? That is where the test sprite I created comes in to "simulate" a scenario. In the first (left) GIF, I ran a simulation with the old implementation, where we can see that the sprite would "reset" itself (intentionally, I may add, given how this *is* a test). Over to the next GIF on the right, I ran another test utilizing the new method, which would "fix" such an oversight.
![Status3Test_OldMethodTest](https://github.com/JackTheSpades/SpriteToolSuperDelux/assets/5531548/311b4cdc-0dde-4fb9-9f64-374b0eca49b1) ![Status3Test_NewMethodTest](https://github.com/JackTheSpades/SpriteToolSuperDelux/assets/5531548/73a64ee3-d564-455a-bce6-2553a6680455)
Naturally, such an instance is overly exaggerated for the sake of demonstration.

My other train of thought while tackling this was future-proofing for if/when PIXI ever allows for inserting Custom Sprites over the Vanilla Slots (I have my own idea on how to implement this, which I have yet to implement since I was focused on these hypotheses). For this, we'll "skip" the Vanilla Sprite check for the Status Handler, which would also apply to "Tweaked" Vanilla Sprites inserted through the Custom Slot. As for what happens when we use this in conjunction with the old method, well, I'll let the results speak for themselves.
![Status3Test_VanillaForceTest](https://github.com/JackTheSpades/SpriteToolSuperDelux/assets/5531548/100fac6e-78d6-4c7b-912c-a443e5a8732d) ![Status3Test_TweakForceTest](https://github.com/JackTheSpades/SpriteToolSuperDelux/assets/5531548/ac941cf3-2288-4f8d-ad96-b0c3f6012f38)
For the "Tweaked" Goomba Sprite (Second/Right GIF), I didn't touch its Main Pointer in Mesen, which turns out to point somewhere that ends in a crash. With the Vanilla Koopa (Sprite 01), I implemented a small hack to make its Main Pointer jump to the actual Stock SMW Handler for Status 8 to see what would happen if I try to "kill" it. This ends up giving us a wonderful sight of it strolling along before it deletes itself from usual Status 3 Code.
Obviously, the above scenarios should (and hopefully will) never happen if we play our cards right, but it's fun to see an amusing what-if event.

Atari2 commented:
Okay now I understand this a bit better, I think this way of trying to fix this is the most reasonable solution, unfortunately, while I was checking for hijack conflicts, I found this:
![image](https://github.com/JackTheSpades/SpriteToolSuperDelux/assets/53220763/16a7ce33-006d-4879-9e5a-8d5b2b252496)
So this means that the hijack you added at $019AFE would conflict with the SA-1 patch, this needs to be fixed.

Also about this:

> My other train of thought while tackling this was future-proofing for if/when PIXI ever allows for inserting Custom Sprites over the Vanilla Slots (I have my own idea on how to implement this, which I have yet to implement since I was focused on these hypotheses)
I have been contacted around 3/4 months ago by @blackmagemathie and she said she wanted to implement such a system to allow for custom sprites over the vanilla slots. We chatted for a bit on discord and the ideas seemed fine, however I haven't heard from her in about a month and a half so I have no idea if it is actually being worked on or not. Just wanted to say this to avoid 2 people from working on the same thing and as such wasting their time.

SubconsciousEye commented:
I have fixed the SA-1 Hijack Conflict by moving it over, as SA-1 only really changes two bytes. I also opted to leave the original hijack location for LoROMs on the off chance some resource may cause mass smushing of multiple sprites in quick succession.

> I have been contacted around 3/4 months ago by @blackmagemathie and she said she wanted to implement such a system to allow for custom sprites over the vanilla slots. We chatted for a bit on discord and the ideas seemed fine, however I haven't heard from her in about a month and a half so I have no idea if it is actually being worked on or not. Just wanted to say this to avoid 2 people from working on the same thing and as such wasting their time.

I'll be honest, I completely forgot that mathie was working on such a feature, given how, as you said, there didn't seem to be much communication since. I had already hypothesized and written most of the code snippets and notes while I was investigating other stuff. I suppose I can wait to see if mathie responds or if we could talk amongst ourselves. My approach would have been more "expanding" upon PIXI, as a carryover from romi's Spritetool, allowing for inserting "Tweaked" Vanilla Sprites through the Custom Slot, which technically *also* means that you could copy exact Vanilla Sprites over, including the **Goal Tape**. I'll be willing to wait or discuss it over more in-depth if any of us so choose.

Atari2 commented:
Most of the communication I had with mathie was in discord dms so maybe that's why it didn't seem like there was no communication since, but actually we talked a lot. Now as I said I haven't heard from her since september so maybe something happened but I can always ask

SubconsciousEye commented:
Understood, I'll leave my DMs open in case something comes up. Hope you have a nice day!
